pkg/virt-controller/watch/snapshot/snapshot.go

```diff
@@ -645,6 +677,85 @@ func (ctrl *VMSnapshotController) updateSnapshotStatus(vmSnapshot *snapshotv1.Vi
 	return nil
 }
 
+func vmsnapshotHasIndication(vmSnapshot *snapshotv1.VirtualMachineSnapshot, indication snapshotv1.Indication) bool {
+	for _, currIndication := range vmSnapshot.Status.Indications {
+		if indication == currIndication {
+			return true
+		}
+	}
+	return false
+}
+
+func (ctrl *VMSnapshotController) checkGuestAgentParticipation(vmSnapshot *snapshotv1.VirtualMachineSnapshot) (bool, error) {
+	condManager := controller.NewVirtualMachineInstanceConditionManager()
+	vm, err := ctrl.getVM(vmSnapshot)
+	if err != nil {
+		return false, err
+	}
+	vmi, exists, err := ctrl.getVMI(vm)
+	if err != nil || !exists {
+		return false, err
+	}
+
+	return condManager.HasCondition(vmi, kubevirtv1.VirtualMachineInstanceAgentConnected), nil
+}
+
+func (ctrl *VMSnapshotController) replaceGuestAgentIndication(vmSnapshot *snapshotv1.VirtualMachineSnapshot) error {
+	for i, indication := range vmSnapshot.Status.Indications {
+		if indication == snapshotv1.VMSnapshotNoGuestAgentIndication {
+			vmSnapshot.Status.Indications[i] = snapshotv1.VMSnapshotGuestAgentIndication
+			break
+		}
+	}
+
+	if _, err := ctrl.Client.VirtualMachineSnapshot(vmSnapshot.Namespace).Update(context.Background(), vmSnapshot, metav1.UpdateOptions{}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ctrl *VMSnapshotController) freezeGuestFSIfNeeded(vmSnapshot *snapshotv1.VirtualMachineSnapshot) error {
+	if !vmsnapshotHasIndication(vmSnapshot, snapshotv1.VMSnapshotOnlineSnapshotIndication) {
+		return nil
+	}
+
+	GAParticipation, err := ctrl.checkGuestAgentParticipation(vmSnapshot)
+	if err != nil || !GAParticipation {
+		return err
+	}
+
+	vmName := vmSnapshot.Spec.Source.Name
+	log.Log.V(3).Infof("Freezing vm %s file system before taking the snapshot", vmName)
+
+	defer timeTrack(time.Now(), "Freezing vm")
```

Comment:
this time span will also include `updateVMSnapshotError` or `replaceGuestAgentIndication`

Reply:
I know.. didnt think its a problem, but ill change it to only wrap the freeze call